src/Core/multiplier.scala

```diff
@@ -17,14 +17,60 @@ class Multiplier(implicit conf: FlexpretConfiguration) extends Module {
     val func   = UInt(INPUT, 4)
     val result = Bits(OUTPUT, 32)
   }
-  
-  val op1 = Mux(io.func === MUL_HU, Cat(Bits(0, 1), io.op1).toSInt, Cat(io.op1(31), io.op1).toSInt)
-  val op2 = Mux(io.func === MUL_HSU || io.func === MUL_HU, Cat(Bits(0, 1), io.op2).toSInt, Cat(io.op2(31), io.op2).toSInt)
-  val mul_result = op1 * op2
-  val result = Mux(io.func === MUL_L, mul_result(31, 0), mul_result(63, 32))
+
+  val op1 = Mux(io.func === MUL_HU || io.func === DIV_LU || io.func === REM_LU, Cat(Bits(0, 1), io.op1).toSInt, Cat(io.op1(31), io.op1).toSInt)
+  val op2 = Mux(io.func === MUL_HSU || io.func === MUL_HU || io.func === DIV_LU || io.func === REM_LU, Cat(Bits(0, 1), io.op2).toSInt, Cat(io.op2(31), io.op2).toSInt)
+  val mul_result = Mux(io.func === REM_L || io.func === REM_LU, rem(op1, op2),
+                      Mux(io.func === DIV_L || io.func === DIV_LU, div(op1, op2), op1 * op2));
+  val result = Mux(io.func === MUL_L || io.func === DIV_L || io.func === DIV_LU || io.func === REM_L || io.func === REM_LU, mul_result(31, 0), mul_result(63, 32))
 
   // 2 cycle
   io.result := Reg(next = result)
 
-}
+  def rem(dividend: Bits, divider: Bits) : Bits = {
+    return divider_signed(dividend, divider, Bool(true));
+  }
+
+  def div(dividend : Bits, divider: Bits) : Bits = {
+    return divider_signed(dividend, divider, Bool(false));
+  }
+
+  def divider_signed(dividend: Bits, divider: Bits, isRemainder: Bool) : Bits = {
+
+    val signed_dividend = dividend(32);
+    val signed_divider = divider(32);
+    val signed_quotient = signed_dividend^signed_divider;
+    val signed_remainder = signed_dividend;
+
+    val udividend = Mux(signed_dividend, ~dividend + Bits(1,1), dividend);
+    val udivider = Mux(signed_divider, ~divider + Bits(1,1), divider);
 
+    var result = divider_unsigned(udividend, udivider, isRemainder);
+
+    return Mux((isRemainder && signed_remainder) || (!isRemainder && signed_quotient),
+      ~result + Bits(1,1),
+      result);
+  }
+
+  def divider_unsigned(dividend : Bits, divider: Bits, isRemainder: Bool) : Bits = {
```

Comment:
I think this implementation could be simplified to use less logic and be more readable. Some observations:
- The subtraction here is 64-bits, but if I remember correctly, iterative integer division can be done here with only a 32-bit subtraction. 
- I don't think quite so many variables need to be saved between iterations of the for loop (`var` outside the loop).

 I would need to code it up to check, and I've haven't written Chisel code in a long time so there's probably an ever better way to do this, but something like this may work:
```
var quotient = Bits(0, 32);
var remainder = Bits(0, 32);

for (i <- 31 to 0) {
    val temp_remainder = Cat(remainder(30, 0), dividend(i));
    val temp_result = temp_remainder - divider;
    val is_positive = ~temp_result(31);
    quotient(i) = is_positive;
    remainder = Mux(is_positive, temp_result, temp_remainder);
}
```

Reply:
We'll probably have to work around it for now since we're still using Chisel2, but once we port to Chisel3 (which I can work on starting next month) there'll be a lot less language warts and we'll be able to get upstream support for Chisel.